backends/aerospike.go

```diff
@@ -13,59 +15,111 @@ import (
+
+	return formatAerospikeError(err, "PUT")
+}
+
+func (db *Aerospike) NewUuidKey(namespace string, key string) (*as.Key, error) {
+	asKey, err := as.NewKey(namespace, setName, key)
+	if err != nil {
+		return nil, formatAerospikeError(err, "NEW_KEY")
+	}
+	return asKey, nil
+}
+
+type AerospikeBackend struct {
 	cfg     config.Aerospike
-	client  *as.Client
+	client  AerospikeDBClient
 	metrics *metrics.Metrics
 }
 
-func NewAerospikeBackend(cfg config.Aerospike, metrics *metrics.Metrics) *Aerospike {
+func NewAerospikeBackend(cfg config.Aerospike, metrics *metrics.Metrics) *AerospikeBackend {
+	if cfg.Host == "" {
+		log.Fatalf("Cannot connect to empty Aerospike host")
+	}
+	if cfg.Port <= 0 {
+		log.Fatalf("Cannot connect to Aerospike host at port %d", cfg.Port)
+	}
 	client, err := as.NewClient(cfg.Host, cfg.Port)
 	if err != nil {
-		log.Fatalf("Error creating Aerospike backend: %v", err)
-		panic("Aerospike failure. This shouldn't happen.")
+		log.Fatalf("Error creating Aerospike backend: %v", formatAerospikeError(err, "NewAerospikeBackend"))
+		panic("AerospikeBackend failure. This shouldn't happen.")
 	}
 	log.Infof("Connected to Aerospike at %s:%d", cfg.Host, cfg.Port)
 
-	return &Aerospike{
+	return &AerospikeBackend{
 		cfg:     cfg,
-		client:  client,
+		client:  &Aerospike{client},
 		metrics: metrics,
 	}
 }
 
-func (a *Aerospike) Get(ctx context.Context, key string) (string, error) {
-	asKey, err := as.NewKey(a.cfg.Namespace, setName, key)
+func (a *AerospikeBackend) Get(ctx context.Context, key string) (string, error) {
+	asKey, err := a.client.NewUuidKey(a.cfg.Namespace, key)
 	if err != nil {
 		return "", err
 	}
-	rec, err := a.client.Get(nil, asKey, "value")
+	rec, err := a.client.Get(asKey)
 	if err != nil {
 		return "", err
 	}
 	if rec == nil {
-		return "", errors.New("client.Get returned a nil record. Is aerospike configured properly?")
+		return "", errors.New("Aerospike GET. Nil record")
 	}
 	a.metrics.RecordExtraTTLSeconds(float64(rec.Expiration))
-	return rec.Bins[binValue].(string), nil
+
+	value, found := rec.Bins[binValue]
+	if !found {
+		return "", errors.New("Aerospike GET. No 'value' bucket found")
+	}
+
+	str, isString := value.(string)
+	if !isString {
+		return "", errors.New("Aerospike GET. Retrieved value is not a string")
+	}
+
+	return str, nil
 }
 
-func (a *Aerospike) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
-	asKey, err := as.NewKey(a.cfg.Namespace, setName, key)
+func (a *AerospikeBackend) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
+	asKey, err := a.client.NewUuidKey(a.cfg.Namespace, setName)
 	if err != nil {
 		return err
 	}
 	if ttlSeconds == 0 {
 		ttlSeconds = a.cfg.DefaultTTL
 	}
-	bins := as.BinMap{
-		binValue: value,
-	}
-	err = a.client.Put(&as.WritePolicy{
-		Expiration: uint32(ttlSeconds),
-	}, asKey, bins)
+	return a.client.Put(asKey, value, ttlSeconds)
+}
+
+func formatAerospikeError(err error, caller string) error {
```

Comment:
Inside this function we can group the different Aerospike DB error codes to whatever code/message we want to use to log them. Something like:
```
118   func formatAerospikeError(err error, caller string) error {
119       if err != nil {
120           if aerr, ok := err.(ase.AerospikeError); ok {
121 -             return fmt.Errorf("%s Aerospike error: %s. Code: %d", caller, aerr.Error(), aerr.ResultCode())
    +              switch aerr.ResultCode() {
    +              case RACK_NOT_DEFINED:
    +                  return fmt.Errorf(" // assign error code and message
    +              case SERVER_NOT_AVAILABLE:
    +                  fallthrough
    +              case SERVER_MEM_ERROR:
    +                  fallthrough
    +              case SERVER_ERROR:
    +                  return fmt.Errorf(" // assign error code and message
    +              case KEY_NOT_FOUND_ERROR:
    +                  return fmt.Errorf(" // assign error code and message
    +              case KEY_EXISTS_ERROR:
    +                  fallthrough
    +              case BIN_EXISTS_ERROR:
    +                  return fmt.Errorf(" // assign error code and message
    +                    .
    +                    .
    +              default:
    +                  return fmt.Errorf("%s Aerospike error: %s. Code: %d", caller, aerr.Error(), aerr.ResultCode())
    +              }
122           }
123       }
124       return err
125   }
backends/aerospike.go
```
What do you think? Should we discuss the possible ways we could group aerospike errors and their error codes and messages? Or should we just print them as they come?